src/pudl/extract/ferc1.py

```diff
@@ -213,429 +208,272 @@
 }
 """A mapping of PUDL DB table names to their XBRL and DBF source table names."""
 
-DBF_TYPES = {
-    "C": sa.String,
-    "D": sa.Date,
-    "F": sa.Float,
-    "I": sa.Integer,
-    "L": sa.Boolean,
-    "M": sa.Text,  # 10 digit .DBT block number, stored as a string...
-    "N": sa.Float,
-    "T": sa.DateTime,
-    "0": sa.Integer,  # based on dbf2sqlite mapping
-    "B": "XXX",  # .DBT block number, binary string
-    "@": "XXX",  # Timestamp... Date = Julian Day, Time is in milliseconds?
-    "+": "XXX",  # Autoincrement (e.g. for IDs)
-    "O": "XXX",  # Double, 8 bytes
-    "G": "XXX",  # OLE 10 digit/byte number of a .DBT block, stored as string
-}
-"""dict: A mapping of DBF field types to SQLAlchemy Column types.
 
-This dictionary maps the strings which are used to denote field types in the DBF objects
-to the corresponding generic SQLAlchemy Column types: These definitions come from a
-combination of the dbfread example program dbf2sqlite and this DBF file format
-documentation page: http://www.dbase.com/KnowledgeBase/int/db7_file_fmt.htm
+class FoxProExtractor:
+    """Generalized class for loading data from foxpro databases into sqlite.
 
-Unmapped types left as 'XXX' which should result in an error if encountered.
-"""
+    When subclassing from this generic extractor, one should implement dataset specific
+    logic in the following manner:
+    1. set DATABASE_NAME. This is going to be used as the file for the resulting sqlite
+       database.
+    2. Overrride get_datastore() method to return the right kind of dataset specific datastore.
 
-PUDL_RIDS: dict[int, str] = {
-    514: "AEP Texas",
-    519: "Upper Michigan Energy Resources Company",
-    522: "Luning Energy Holdings LLC, Invenergy Investments",
-    529: "Tri-State Generation and Transmission Association",
-    531: "Basin Electric Power Cooperative",
-}
-"""Missing FERC 1 Respondent IDs for which we have identified the respondent."""
+    Dataset specific logic and transformations can be injected by overriding:
+    1. finalize_schema() in order to modify sqlite schema. This is called just before the
+       schema is written into the sqlite database. This is good place for adding primary and/or
+       foreign key constraints to tables.
+    2. transform_table(table_name, df) will be invoked after dataframe is loaded from the foxpro
+       database and before it's written to sqlite. This is good place for table-specific
+       preprocessing and/or cleanup.
+    3. postprocess() is called after data is written to sqlite. This can be used for database
```

Comment:
`postprocess()` would be database-level in the case of the missing `respondent_id` values because we have to actually scan the whole DB to know which IDs are missing, right?

Reply:
Correct.